pkg/apis/resource/types.go

```diff
@@ -37,90 +307,335 @@ type ResourceClaim struct {
 	// +optional
 	metav1.ObjectMeta
 
-	// Spec describes the desired attributes of a resource that then needs
-	// to be allocated. It can only be set once when creating the
-	// ResourceClaim.
+	// Spec describes what is being requested and how to configure it.
+	// The spec is immutable.
 	Spec ResourceClaimSpec
 
-	// Status describes whether the resource is available and with which
-	// attributes.
+	// Status describes whether the claim is ready to use and what has been allocated.
 	// +optional
 	Status ResourceClaimStatus
 }
 
-// ResourceClaimSpec defines how a resource is to be allocated.
+// ResourceClaimSpec defines what is being requested in a ResourceClaim and how to configure it.
 type ResourceClaimSpec struct {
-	// ResourceClassName references the driver and additional parameters
-	// via the name of a ResourceClass that was created as part of the
-	// driver deployment.
-	ResourceClassName string
+	// Devices defines how to request devices.
+	//
+	// +optional
+	Devices DeviceClaim
 
-	// ParametersRef references a separate object with arbitrary parameters
-	// that will be used by the driver when allocating a resource for the
-	// claim.
+	// Controller is the name of the DRA driver that is meant
+	// to handle allocation of this claim. If empty, allocation is handled
+	// by the scheduler while scheduling a pod.
+	//
+	// Must be a DNS subdomain and should end with a DNS domain owned by the
+	// vendor of the driver.
+	//
+	// This is an alpha field and requires enabling the DRAControlPlaneController
+	// feature gate.
 	//
-	// The object must be in the same namespace as the ResourceClaim.
 	// +optional
-	ParametersRef *ResourceClaimParametersReference
+	Controller string
+}
 
-	// Allocation can start immediately or when a Pod wants to use the
-	// resource. "WaitForFirstConsumer" is the default.
+// DeviceClaim defines how to request devices with a ResourceClaim.
+type DeviceClaim struct {
+	// Requests represent individual requests for distinct devices which
+	// must all be satisfied. If empty, nothing needs to be allocated.
+	//
 	// +optional
-	AllocationMode AllocationMode
+	// +listType=atomic
+	Requests []DeviceRequest
+
+	// These constraints must be satisfied by the set of devices that get
+	// allocated for the claim.
+	//
+	// +optional
+	// +listType=atomic
+	Constraints []DeviceConstraint
+
+	// This field holds configuration for multiple potential drivers which
+	// could satisfy requests in this claim. It is ignored while allocating
+	// the claim.
+	//
+	// +optional
+	// +listType=atomic
+	Config []DeviceClaimConfiguration
```

Comment:
Should we be consistent in the naming of the types here? Why `DeviceRequest` and `DeviceConstraint`, but then `DeviceClaimConfiguration`? Should they all have `Claim` in the name (or not have it, for that matter)?

Reply:
The `Claim` here is needed because the different `*Configuration` structs are slightly different.

Reply:
I read the question as "why not DeviceClaimConstraint and DeviceClaimRequest".  IMO those would be more verbose for no real gain.  Claim config and class config are different, as Patrick said.